### ISSUE TITLE
ppc64el architecture problem

### ISSUE THREAD

**Issue by razorkam** [FreeTDS/freetds#286]
Hi guys. I have quite a strange problem running FreeTDS.
Well, from the very beginning: 
There are two servers, both are powered by Ubuntu 18.04 LTS. The only difference is : first one has **x86_64** arch ( Intel CPU ) and the second one has **ppc64le** arch  ( IBM Power9 CPU ).

Both machines have packages **tdsodbc** and **freetds-bin** installed ( ODBC TDS driver and binary FreeTDS utils ). Package's FreeTDS versions are the same on both machines : **1.00.82-2**. 

I'm testing FreeTDS using **tsql** utility connecting to **Microsoft SQL Server 2017 (RTM) - 14.0.1000.169 (X64)** running on the third server under Windows 10 ( if it could make sense to detect the problem ). In file **freetds.conf** _tds version=auto_ is set. 

In case of testing on x86_64 arch everything is OK, TDS protocol version was automatically selected as 7.3, all queries to database via **tsql** was working great. 
The problem is under ppc64le arch. When I tried get connected the same way, TDS protocol version was detected as 4.2! That's why I faced some problems, for example: Unicode usage was unavailable, but I strongly need it.

I tried to set **tds version = 7.3** manually in freetds.conf. But got following error: 

> Error 20017 (severity 9):
> 	Unexpected EOF from the server
> Error 20002 (severity 9):
> 	Adaptive Server connection failed

Looks like server doesn't support TDS 7.3, but I know it does ( by testing on x86_64 server ).
There are nothing clear too in freetds.log with full debug flags  :

> 14:55:00.458440 28791 (login.c:954):quietly sending TDS 7+ login packet
> 14:55:00.458509 28791 (token.c:305):tds_process_login_tokens()
> 14:55:00.472992 28791 (util.c:165):Changed query state from IDLE to DEAD
> 14:55:00.473005 28791 (util.c:322):tdserror(0x9ec46a70c90, 0x9ec46a70cc0, 20017, 0)
> 14:55:00.473025 28791 (util.c:352):tdserror: client library returned TDS_INT_CANCEL(2)
> 14:55:00.473034 28791 (util.c:375):tdserror: returning TDS_INT_CANCEL(2)
> 14:55:00.473044 28791 (token.c:313):looking for login token, got  0()
> 14:55:00.473058 28791 (token.c:132):tds_process_default_tokens() marker is 0()
> 14:55:00.473071 28791 (token.c:135):leaving tds_process_default_tokens() connection dead
> 14:55:00.473085 28791 (login.c:534):login packet accepted
> 14:55:00.473098 28791 (util.c:322):tdserror(0x9ec46a70c90, 0x9ec46a70cc0, 20002, 0)
> 14:55:00.473116 28791 (util.c:352):tdserror: client library returned TDS_INT_CANCEL(2)
> 14:55:00.473126 28791 (util.c:375):tdserror: returning TDS_INT_CANCEL(2)
> 14:55:00.473137 28791 (mem.c:653):tds_free_all_results()

Thanks for your attention, do you have any ideas how to deal with it , or how to debug it?

P.S. Should I ask this question using the mailing list? If so, could you please give me subscription to razorkam@gmail.com ?

**Comment by razorkam** [comment]
Another very interesting feature has been found: **isql** utility from unixODBC works fine - queries and returns Unicode data correctly. 
But if I try to do query from my application code via unixODBC, it works the wrong way - the same like **tsql** does.

**Comment by freddy77** [comment]
The isql inconsistency is weird.
I suspect that ppc little endian is not correctly supported. Can you try this patch?

```
--- a/include/freetds/bytes.h
+++ b/include/freetds/bytes.h
@@ -199,7 +199,7 @@ typedef union {
    uint32_t _tds_i = bswap_32(val); TDS_PUT_UA4LE(ptr,_tds_i); } while(0)
 #endif
 
-#if defined(__GNUC__) && defined(__powerpc__)
+#if defined(__GNUC__) && defined(__powerpc__) && defined(WORDS_BIGENDIAN)
 # undef TDS_GET_UA2LE
 # undef TDS_GET_UA4LE
 static inline uint16_t

```

**Comment by razorkam** [comment]
@freddy77  Thanks for you answer. 
I've just built FreeTDS using your patch. Definitely, it's getting better! Now I can connect using TDS protocol v 7.4 - it's OK. 

But encoding problems are still presented. Look at this **tsql** output, please. 

> locale is "C.UTF-8"
> locale charset is "UTF-8"
> using default charset "UTF-8"
> 1> select * from DIFF_data.dbo.test_table
> 2> go
> id	pid	title
> 1	NULL	Ðîññèÿ
> 10	1	Öåíòðàëüíûé ðåãèîí
> 20	1	Çàïàäíàÿ ñèáèðü
> 30	1	Âîñòî÷íàÿ ñèáèðü
> 111	10	Ìîñêâà
> 222	10	Ñàíêò Ïåòåðáóðã
> 333	10	Íèæíèé íîâãîðîä
> 111	30	Èðêóòñê
> 222	30	Âëàäèâîñòîê
> 111	20	Åêàòåðèíáóðã
> 222	20	Òþìåíü
> (11 rows affected)
> 1> exit
> 

Here we have 3 columns of query result, and the third one ( named  `title` ) contains Cyrillic symbols in database.
 I was expecting to get it as UTF-8 ( as I was requesting , 'using default charset "UTF-8"' ) in readable format, but seeing a broken WINDOWS-1251 encoding ( that's weird ). 

So, to be clear: for example, running the same query on x86_64 I  can see correct string "Центральный регион" instead of "Öåíòðàëüíûé ðåãèîí". Note that symbols here is presented, but ill-formed. I case of using TDS v. 4.2 without your patch all symbols has no information at all ( which looks like "???????????????" ).

**Comment by razorkam** [comment]
Has this in freetds.log:

> 17:35:29.028818 20752 (iconv.c:785):setting server single-byte charset to "CP1251"
> 17:35:29.028829 20752 (iconv.c:444):Charset 14 not supported by iconv, using "ISO-8859-1" instead

Is it FreeTDS problem or iconv's? Any ideas how to fix it?

**Comment by razorkam** [comment]
Yeah, finally fixed it. It was my iconv problem ( it was not detected during  project configuring stage, AM_ICONV macro was missing ). Now it's ok. Thank you very much. 

Do you plan to include this patch to your nearest release?

**Comment by freddy77** [comment]
Patch merged to 1.0, 1.1 and master branch. Opened a bug for Ubuntu at https://bugs.launchpad.net/ubuntu/+source/freetds/+bug/1826851. For 1.0 file at http://www.freetds.org/files/stable/freetds-1.00.112.tar.bz2.

**Comment by freddy77** [comment]
Closing, see reply in bug, thanks Steve
